test/e2e/common/k8s.py

```diff
@@ -0,0 +1,199 @@
+        reference: CustomResourceReference, custom_resource: dict):
+    _api_client = _get_k8s_api_client()
+    _api = client.CustomObjectsApi(_api_client)
+
+    if reference.namespace is None:
+        return _api.create_cluster_custom_object(
+            reference.group, reference.version, reference.plural, custom_resource)
+    return _api.create_namespaced_custom_object(
+        reference.group, reference.version, reference.namespace, reference.plural, custom_resource)
+
+
+def delete_custom_resource(reference: CustomResourceReference):
+    _api_client = _get_k8s_api_client()
+    _api = client.CustomObjectsApi(_api_client)
+    if reference.namespace is None:
+        return _api.delete_cluster_custom_object(
+            reference.group, reference.version, reference.plural, reference.name)
+    return _api.delete_namespaced_custom_object(
+        reference.group, reference.version, reference.namespace, reference.plural, reference.name)
+
+
+def get_resource(reference: CustomResourceReference):
+    _api_client = _get_k8s_api_client()
+    _api = client.CustomObjectsApi(_api_client)
+
+    if reference.namespace is None:
+        return _api.get_cluster_custom_object(
+            reference.group, reference.version, reference.plural, reference.name)
+
+    return _api.get_namespaced_custom_object(
+        reference.group, reference.version, reference.namespace, reference.plural, reference.name)
+
+
+def get_resource_exists(resource: CustomResourceReference) -> bool:
+    return get_resource(resource) is not None
+
+
+def wait_resource_consumed_by_controller(
+        reference: CustomResourceReference, wait_periods: int = 3, period_length: int = 10):
+    if not get_resource_exists(reference):
+        logging.error(f"Resource {reference} does not exist")
+        return None
+
+    for _ in range(wait_periods):
+        resource = get_resource(reference)
+
+        if 'status' in resource:
+            return resource
+
+        sleep(period_length)
+
+    logging.error(
+        f"Wait for resource {reference} to be consumed by controller timed out")
+    return None
+
+
+def wait_resource_synced(reference: CustomResourceReference,
+                         wait_periods: int = 2, period_length: int = 60):
+    if not get_resource_exists(reference):
+        logging.error(f"Resource {reference} does not exist")
+        return False
+
+    if wait_resource_consumed_by_controller(reference) is None:
+        return False
+
+    for _ in range(wait_periods):
+        logging.debug(f"Waiting for resource {reference} to be synced")
+
+        reference = get_resource(reference)
+        if 'status' not in reference or 'conditions' not in reference['status']:
+            logging.error(
+                f"Expected .status.conditions property to exist in {reference}")
+            return False
+
+        conditions: Dict = reference['status']['conditions']
+        if 'ACK' not in conditions or 'ResourceSynced' not in conditions['ACK']:
```

Comment:
Ohhh actually this line of code is a result of a faulty refactoring. The parameter used to be called `resource` rather than `reference`, but then I was reusing the variable name. When I renamed `resource` to `reference` it modified the local variable as well (on line 143):
https://github.com/aws/aws-controllers-k8s/pull/587/files/cc33198c1102ad5dc100a4d5706fc5d2ce69c8e7#diff-c17786fb0613026f7c25ee4300969ab7ed42e3e6b219d5f871345cf7aefa9af5R143

So actually, this process of splitting up the status fields by path name is from the K8s API (or the Python interpretation) itself, nothing to do with `dataclass` (pitchforks can be lowered again).

Reply:
Nevertheless I will provide a helper method to pull out both this and `ACK.Terminal`, just to break up the code flow. (And I've renamed the variables to make sense again)